### PR TITLE
[Actions] Label PRs where the formula is `bottle :unneeded`

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] ~Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?~
- [x] ~Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?~
- [x] ~Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?~

-----

- The vast majority of formulae here have bottles, but there are some that we can merge through the GitHub web UI buttons.
- Yes, we'll still check the diff of submitted formulae, but for those short on time, it might be good to know that we can review submissions and merge them on GitHub without having to be near a computer to pull the bottles.
- This was also a nice look into how we can potentially in future auto-label other conditions for formulae (new formula, test failure, etc).

### CODE REVIEW

**Comment by SMillerDev** [comment]
Looks good to me

**Comment by fxcoudert** [comment]
If I read this correctly (Ruby is not my forte), this will apply the label to any PR that has at least one `bottle :unneeded`. But we need it to apply to PRs where all formulas are bottle uneeded.

**Comment by issyl0** [comment]
> If I read this correctly (Ruby is not my forte), this will apply the label to any PR that has at least one `bottle :unneeded`. But we need it to apply to PRs where all formulas are bottle unneeded.

Oops - I didn't do enough testing of all the different possibilities. :flushed: I've fixed this now, and in doing so made it more extensible for us to maybe add other label checks in the future.

File: .github/workflows/labels.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+jobs:
+  label:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-ruby@v1
+        with:
+          ruby-version: '2.6.3'
+      - run: gem install octokit
+      - run: ruby .github/workflows/scripts/add-bottle-unneeded-label.rb
```
Comment (issyl0) [inline .github/workflows/labels.yml]: Should we name this script to be more generic? Maybe `add-labels`?

Comment (issyl0) [reply .github/workflows/labels.yml]: I did this. This should be good to go now!

File: .github/workflows/scripts/add-labels.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+client = Octokit::Client.new(access_token: ENV["GITHUB_TOKEN"])
+
+repo = ENV["GITHUB_REPOSITORY"]
+pr_number = ENV["GITHUB_REF"].split("/")[2]
+
+pr_labels = {
+  no_bottles: 0
+}
+
+client.pull_request_files(repo, pr_number).each do |file|
```
Comment (dawidd6) [inline .github/workflows/scripts/add-labels.rb]: This function is called twice I see. Might be a good idea to call it once, always lesser GitHub API knocks.

Comment (issyl0) [reply .github/workflows/scripts/add-labels.rb]: I think it'll work with just one call. Fixed.

**Review by dawidd6 — APPROVED** [review]

File: .github/workflows/scripts/add-labels.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+pr_files.each do |file|
+  formula = file[:filename]
+
+  if File.exist?(formula) && File.read(formula).match(/bottle :unneeded/)
+    pr_labels[:no_bottles] += 1
+  end
+end
+
+if pr_files.count == pr_labels[:no_bottles]
```
Comment (bayandin) [inline .github/workflows/scripts/add-labels.rb]: Will this work as expected if we have deleted files/formulae in a PR? And the same question about the case when we update a formula which has a versioned analogue (when PR contains alias changing)?

**Comment by bayandin** [comment]
I like it a lot, thanks @issyl0 👍 

Can we add tests for `add-labels.rb`, just to be sure that all known cases are working fine?

PS It seems my question landed just after it was merged :)

File: .github/workflows/scripts/add-labels.rb (hunk 3 above)
Comment (issyl0) [reply .github/workflows/scripts/add-labels.rb]: It doesn't add labels if a formula has been deleted, thanks to the `File.exist?` check.

Comment (issyl0) [reply .github/workflows/scripts/add-labels.rb]: However, you raise a good point about the aliases, that was one thing I didn't try [over in my testing repo](https://github.com/issyl0/homebrew-actions-testing/). I'll give it a go and push a fix if it doesn't work!

**Comment by issyl0** [comment]
Looks like I hit merge a little too soon. Thanks for the feedback.

> Can we add tests for `add-labels.rb`, just to be sure that all known cases are working fine?

I'd be up for this. But where would they live? I guess a spec file in the `scripts` dir for people to run locally? I will probably need some help though, as I've done a lot of tests for Rails-based things before, but not much on plain Ruby scripts like this.

File: .github/workflows/scripts/add-labels.rb (hunk 3 above)
Comment (issyl0) [reply .github/workflows/scripts/add-labels.rb]: This action will [only trigger on changes to `Formula/*.rb`](https://github.com/Homebrew/homebrew-core/pull/48911/files#diff-b59ac451da8b4fc7c34118b14fb4d067R5-R6), so I think we'll be fine for aliases? What behaviour were you expecting?

Comment (bayandin) [reply .github/workflows/scripts/add-labels.rb]: I'm talking about PRs like https://github.com/Homebrew/homebrew-core/pull/48823 (but with `bottle :unneeded`), I'm not sure, but `pr_files.count` will be 2 in this case, but `pr_labels[:no_bottles]` is only 1 (I haven't checked it, just assuming)

Sorry, I poorly formulated the question 🙂

**Comment by bayandin** [comment]
> I'd be up for this. But where would they live? I guess a spec file in the `scripts` dir for people to run locally?

I was thinking about extracting DSL part into https://github.com/Homebrew/brew (or https://github.com/Homebrew/.github). This extracted part should parse PR changes and  provide methods like `formula_updated?`, `revision_bumped?`, `uses_bottle?`, `formula_deleted?` and so on, we could test it as much as we can.
Then a script for assigning labels in `homebrew-core` will be dead simple.

Just a thought.. didn't think much about it.

File: .github/workflows/scripts/add-labels.rb (hunk 3 above)
Comment (issyl0) [reply .github/workflows/scripts/add-labels.rb]: Thanks - I understand the problem now! I've fixed this in #48919.

**Pull request merged**